Deprecate internal interpolation in `InflationIndex` and `ZeroInflationIndex`

ralfkonrad opened the pull request:
I would like to work on #1050.

lballabio commented:
Thanks for picking this up, Ralf.  Before you do more work on this: I'm not sure this also applies to year-on-year inflation indexes, so you might leave them alone for now.  They might be the subject of a whole other discussion for another time.

lballabio commented on ql/indexes/inflation/aucpi.hpp:
```diff
@@ -2,6 +2,7 @@
 
 /*
  Copyright (C) 2009 Chris Kenyon
+ Copyright (C) 2007 Ralf Konrad Eckel
```
2021 (in the other files, too)

ralfkonrad replied:
👍  Copy & MisPasted.

ralfkonrad commented:
> Thanks for picking this up, Ralf. Before you do more work on this: I'm not sure this also applies to year-on-year inflation indexes, so you might leave them alone for now. They might be the subject of a whole other discussion for another time.

Should I roll back the changes for the YoY than:
- Do not deprecate the constructors? 
- Remove or leave constructor without `bool interpolated`? 
- Will the YoY stick with `bool interpolated_;` and `bool interpolated() const;` later on?

Besides that we should discuss how proceed from here.

Are you finished with modernizing `c++11`? The modernizing touches a lot of files so I see a risk to run into merge conflicts here as the `clang-format` mixes up the files as well. I might postpone further changes.

I would be in favor to do the changes in `ql/indexes` and the corresponding `testsuite` cases first. By setting the macro
```cpp
#define QL_TO_BE_DEPRECATED_INTERPOLATED_INFLATION_INDEXES                    \
    /*! \deprecated The use of interpolated is not a feature of the index but \
     *              of the coupons etc. using the index.                      \
     *              Use the other constructor instead.                        \
     *              Deprecated in version 1.22. */                            \
    //QL_DEPRECATED
```
one could unset the deprecation, avoid the warning and errors and the changes than could already be merged.

Based on that I could do in a second step work on the coupons etc.

ralfkonrad commented:
Luigi, does it break the existing interface if I make `bool InflationIndex::interpolated() const;` virtual?

At some point we have to deprecate `interpolate()` as well and this would allow it to override and to deprecate it for `ZeroInflationIndex` and `YoYInflationIndex` separately. That makes it a lot easier to work on them.

Or should I just add `QL_DEPRECATED bool ZeroInflationIndex::interpolated() const { return interpolated_; }` without virtual in `InflationIndex`?

lballabio commented:
I'd roll the YoY indexes back to the way they were.   Later on, I'd think about whether we should keep modeling year-on-year inflation this way.  Again, there's no YOYHICP index in the real world; there is HICP, and there are coupons that pay its year-on-year performance.

I _think_ you can deprecate `interpolated()` in the base class and redefine it, not deprecated, in `YoYInflationIndex`? I'm not sure it will work, though.

lballabio commented:
Good point about merging in stages.  However, the new constructors give you indexes that wouldn't work if the coupons don't manage the interpolation already.  If you want to do this in two steps, you might go the other way around: that is, leave this branch for the moment and open another one in which you work on the coupon instead.  The new coupon code can ignore the `interpolated` method of the index.  Once that is merged, you can come back here and deprecate the method; at that point, you won't need the `QL_TO_BE_DEPRECATED_INTERPOLATED_INFLATION_INDEXES` macro since you'll be able to replace the deprecated calls with the new ones.

lballabio commented:
Oh, and yes, I'm done with the automated part of the C++11 modernization.  I will change something else, but those will be manual changes and won't touch nearly as many files.

ralfkonrad commented:
> I'd roll the YoY indexes back to the way they were. Later on, I'd think about whether we should keep modeling year-on-year inflation this way. Again, there's no YOYHICP index in the real world; there is HICP, and there are coupons that pay its year-on-year performance.

Understood, I have reverted the YoY changes.

_My naive(?) understanding of YoY is that it is something like `HICP(t)/HICP(t-12M)` therefore I thought there is also only one fixing per month._
 
> I _think_ you can deprecate `interpolated()` in the base class and redefine it, not deprecated, in `YoYInflationIndex`? I'm not sure it will work, though.

Works at least with `VS C++`. Let's see if this works with the other `C++` dialects as well.

ralfkonrad commented:
> However, the new constructors give you indexes that wouldn't work if the coupons don't manage the interpolation already.

First I thought that's a good idea. But don't we have the same problem the other way around than:

```cpp
/* old style, deprecated, but still okay */
InflationCoupon(/* interpolated, */ InflationIndex(interpolated));

/* new style, okay */
InflationCoupon(interpolated, InflationIndex(/* interpolated */));

/* old style coupon, new style index
 * could be treated as interpolated = false */
InflationCoupon(/* interpolated, */ InflationIndex(/* interpolated */));

/* new style coupon, old style index
 * how does the interpolation now?
 * what do do if the do not match? */
InflationCoupon(interpolated, InflationIndex(interpolated));

```

I would prevent the last two mixed cases and let them `QL_FAIL(...)` when the `InflationIndex(...)` constructor is called in mixed state. Especially the last one does not make sense to me!?

@lballabio Do you agree on this approach?

This means we cannot really separate the PR anyhow. We are moving the `bool interpolated` flag from index to coupon and any intermediate state does not make sense.

If you are not preparing automated `clang tidy` fixes anymore I should be able to merge your fixes manually if needed.

lballabio commented:
We might have to manage intermediate states anyway, as we're deprecating the old-style indexes but not removing them.  I'd tend toward being permissive, so treating your third case as false, as you say, and possibly having the coupon setting win in your fourth case.

Preventing the fourth case would also make sense, but it would require additional effort: the coupon can't distinguish the `InflationIndex()` vs `InflationIndex(interpolated)` case using the current public interface of the index, and I'm not sure I'd add another method for that which would already start deprecated.

ralfkonrad commented:
> Preventing the fourth case would also make sense, but it would require additional effort: the coupon can't distinguish the `InflationIndex()` vs `InflationIndex(interpolated)` case using the current public interface of the index, and I'm not sure I'd add another method for that which would already start deprecated.

I agree on the "current public interface of the index". I had something like
```cpp
class InflationIndex {
    friend bool detail::isNewStyleInflationIndex(InflationIndex index);
    ...
  private:
    bool isNewStyle_;
}

namespace detail {
    bool isNewStyleInflationIndex(InflationIndex index) { return index.isNewStyle_; }
}
```
in mind so the InflationCoupon is able to check via the free function if the styles fit together. Also we do not touch the public interface but of course still have to deprecate the free function later on. This would allow us to be restrictive in both mixed cases.

I am in favor of being restrictive as I do not have to implement four possible "styles" where three of them are deprected already and have to be carefully removed afterwards.

But being permissive would be okayish for me as well. I think we should be consistent here: either allow or fail both mixed styles.

ralfkonrad commented:
Oh, I think `CPICoupon`-wise there is probably less effort than I expected.

In `cpicoupon.hpp` we have already
```cpp
struct CPI {
    //! when you observe an index, how do you interpolate between fixings?
    enum InterpolationType {
        AsIndex,   //!< same interpolation as index
        Flat,      //!< flat from previous fixing
        Linear     //!< linearly between bracketing fixings
    };
};
```
and
```cpp
CPICoupon(Real baseCPI, // user provided, could be arbitrary
          const Date& paymentDate,
          Real nominal,
          const Date& startDate,
          const Date& endDate,
          Natural fixingDays,
          const ext::shared_ptr<ZeroInflationIndex>& index,
          const Period& observationLag,
          CPI::InterpolationType observationInterpolation,
          const DayCounter& dayCounter,
          Real fixedRate, // aka gearing
          Spread spread = 0.0,
          const Date& refPeriodStart = Date(),
          const Date& refPeriodEnd = Date(),
          const Date& exCouponDate = Date());
```
so the `CPI::InterpolationType observationInterpolation` takes care of the mixed styles and the appropriate interpolation already. I think `CPI::AsIndex` can be deprecated as it will become the same as `CPI::Flat` when the index no longer does the interpolation.

Besides that there are a couple of other places which needs attention as they rely on `ZeroInflationIndex::interpolated()`:
- `CPICapFloor`
- `ZeroCouponInflationSwap`
- `ZeroCouponInflationSwapHelper`
- `CPICapFloorTermPriceSurface`

Here we can deprecate some constructors and provide new ones including the `CPI::InterpolationType observationInterpolation` so the `ZeroInflationIndex::interpolated()` call can be replaced by using `observationInterpolation`.

However I am not sure I can avoid using `QL_DEPRECATED ZeroInflationIndex::interpolated()` throughout these classes. At least they need some `QL_DEPRECATED` as well and then of course all the test cases relying on them...

I think the `gcc` compile fails even if one uses a deprecated method only within another deprected method or constructor with `-Werror=deprected`.

_What I wanted to say_: Perhaps it is easier to not deprecate `(Zero)InflationIndex::interpolated()` at this stage but only the constructor. When the constructor finally is removed the `ZeroInflationIndex::interpolated()` always returns `false` and the references in the above classes can be fixed or replaced.

lballabio commented:
> I am in favor of being restrictive as I do not have to implement four possible "styles" where three of them are deprected already and have to be carefully removed afterwards.

Sure — My impression, though, was that making it permissive would result in _less_ code to write.  If the new constructor sets `interpolated_` to false, case 3 becomes the same as case 1; and if we give priority to the coupon, case 4 becomes the same as case 2.  You won't need `isNewStyleInflationIndex`, because the coupon would do the same thing.

lballabio commented:
About deprecation warnings: you can avoid them by surrounding the call to the deprecated method with

```
#if defined(__GNUC__)
#pragma GCC diagnostic push
#pragma GCC diagnostic ignored "-Wdeprecated-declarations"
#endif
#if defined(__clang__)
#pragma clang diagnostic push
#pragma clang diagnostic ignored "-Wdeprecated-declarations"
#endif
#if defined(QL_PATCH_MSVC)
#pragma warning(push)
#pragma warning(disable:4996)
#endif

...code that calls the deprecated method...

#if defined(QL_PATCH_MSVC)
#pragma warning(pop)
#endif
#if defined(__clang__)
#pragma clang diagnostic pop
#endif
#if defined(__GNUC__)
#pragma GCC diagnostic pop
#endif
```

but yes, it might be simpler to deprecate the constructor now and the method later.

ralfkonrad commented:
> it might be simpler to deprecate the constructor now and the method later.

But that postpones the final deprecation of `InflationIndex::interpolated()` till v1.26 or v1.28 if you keep the old style constructor for two or three more release?!

> About deprecation warnings: you can avoid them by surrounding the call to the deprecated method with

I had the same idea. I am trying to combine them into a macro like
```cpp
QL_DEPRECATED_SKIP_WARNING_BEGIN

...code that calls the deprecated method...

QL_DEPRECATED_SKIP_WARNING_END
```
to avoid the spaghetti code if possible. I need this in many places still, especially in the test cases. If this is possible I can try to go for the big bang approach.

> Sure — My impression, though, was that making it permissive would result in less code to write. If the new constructor sets interpolated_ to false, case 3 becomes the same as case 1; and if we give priority to the coupon, case 4 becomes the same as case 2. You won't need isNewStyleInflationIndex, because the coupon would do the same thing.

As above: The `CPICoupon` overrules the `InflationIndex::interpolated()` already using `CPI::InterpolationType observationInterpolation` and I think that is the only non-abstract cashflow using the `ZeroInflationIndex`. So here we do not have mixed styles in the end.

ralfkonrad commented:
Luigi, [my Mac OS `boost` build fails, the `std` succeeds](https://github.com/ralfkonrad/QuantLib/actions/runs/590230804) because `boost::make_shared` complains about the deprecation of the constructor while the `std::make_shared` doesn't.

The line of code that fails is [here](https://github.com/ralfkonrad/QuantLib/blob/1d6330ea4e80e962a21e767eb11089dc3a1b9557/test-suite/inflationcpibond.cpp#L113).

The macro to avoid the deprecated warning seems to work otherwise the std::make_shared should fail as well.

Do you know if this a known issue about boost::make_shared and deprecated constructors? I cannot anything on Google about it.

ralfkonrad commented:
And the [boost based cmake-macos](https://github.com/ralfkonrad/QuantLib/actions/runs/590230906) workflow also builds?!

ralfkonrad commented:
@lballabio: I am at the point to deprecate `bool InflationIndex::interpolated() const;`

This is closely link to
https://github.com/lballabio/QuantLib/blob/c13bff9bda56a1fc673d3d6e4945c1f80cdf1d22/ql/cashflows/cpicoupon.hpp#L34-L41

For example there is code like

https://github.com/lballabio/QuantLib/blob/c13bff9bda56a1fc673d3d6e4945c1f80cdf1d22/ql/instruments/cpicapfloor.cpp#L57-L72

When the index fixings cannot be interpolated anymore, `CPI::AsIndex` becomes effectively `CPI::Flat`.

First I thought I can `QL_DEPRECATED AsIndex` only but that is not allowed in `c++`.

We could `QL_DEPRECATED enum InterpolationType { ... };` totally and replace it with something like

```cpp
struct CPIInterpolation {
    //! when you observe an index, how do you interpolate between fixings?
    enum InterpolationType {
        Flat,      //!< flat from previous fixing
        Linear     //!< linearly between bracketing fixings
    };
};
```
This is probably the most accurate way to do it.

However `struct CPI { enum InterpolationType { ... } }` is heavily used throughout the Inflation implementation and would mean a lot of more deprecated constructors and `CPI::InterpolationType interpolation() const;` interfaces exposing its value later on as most of them currently use it to distinguish the interpolation types (I get 175 warnings at the moment within the library itself not taking into account the testsuite and examples).

So this will be a lot of work which is also error prune. I am not in favor doing this right away.

My approach would be when `bool InflationIndex::interpolated() const;` will be removed to set
```cpp
enum InterpolationType {
    AsIndex = 0, //!< same interpolation as index
    Flat    = 0, //!< flat from previous fixing
    Linear  = 1  //!< linearly between bracketing fixings
};
```
and remove `CPI::AsIndex` internally and later on remove it from `enum InterpolationType` silently in the sense that we cannot deprecate it but only announce it.

Alternatively one could be "nasty" and
```cpp
QL_REQUIRE(cpiInterpolation != CPI::AsIndex, 
           "CPI::AsIndex is no longer supported as the index fixing is always flat.");
```
where appropriate.

What are your thoughts on this? Are their better ideas/approaches?

ralfkonrad commented:
_The (perhaps) best ideas are always there when I have finished the comment._

We remove `CPI::AsIndex` but give the user the possibiliy to re-enable it:
```cpp
    struct CPI {
        //! when you observe an index, how do you interpolate between fixings?
        enum InterpolationType {
#ifdef QL_REENABLE_CPI_ASINDEX
            AsIndex, //!< same interpolation as index
#endif
            Flat,  //!< flat from previous fixing
            Linear //!< linearly between bracketing fixings
        };
    };
```

lballabio commented:
Sorry, I had a few busy days.

The `boost::make_shared` thing _is_ weird.  No, I don't have a clue about it either...

ralfkonrad commented:
> Sorry, I had a few busy days.

No problem.

> The `boost::make_shared` thing _is_ weird. No, I don't have a clue about it either...

First I thought (or was afraid) it might be a problem of the #1055 macro.

The good thing about the weird `boost::make_shared` is the easy workaround `boost::shared_ptr<T>(new T(...));`

ralfkonrad commented:
@lballabio Two questions for the moment:

**Regarding `InflationTermStructure`:**

```cpp
class InflationTermStructure {
...
    virtual bool indexIsInterpolated() const;
    bool indexIsInterpolated_;
...
}
```
incl. the constructor passing the parameter are more or less the equivalents of the `bool InflationIndex::interpolated() const;` etc. They make sense for `YoY` but for the `ZeroInflationTermStructure` they will be effectively become `false` ones `InflationIndex::interpolated()` being removed.

Should we deprecate them as well and redefine them for the `YoY` case like I did for the index?

**Regarding `ZeroCouponInflationSwap`:**

The `ZeroCouponInflationSwap` will loose its capability to use interpolated fixings. First I thought I might extend the class by

```cpp
ZeroCouponInflationSwap(Type type,
                        ...
                        const ext::shared_ptr<ZeroInflationIndex>& infIndex,
                        const bool& useInterpolatedFixings,
                        const Period& observationLag,
                        ...);

bool useInterpolatedFixings_;
```
However this will not work directly as the `ZeroCouponInflationSwap` "only" defines an `IndexedCashFlow`

https://github.com/lballabio/QuantLib/blob/b0c1cdf46a9df8356e498dad4be6828bddf2344f/ql/instruments/zerocouponinflationswap.cpp#L94-L95

and its amount is calculated via `ZeroInflationIndex::fixing();` which will be flat after deprecated interpolated indexes:

https://github.com/lballabio/QuantLib/blob/b0c1cdf46a9df8356e498dad4be6828bddf2344f/ql/cashflows/indexedcashflow.cpp#L25-L33

There is a testcase

https://github.com/lballabio/QuantLib/blob/b0c1cdf46a9df8356e498dad4be6828bddf2344f/test-suite/inflationcpiswap.cpp#L383-L384

which luckily uses `interpolated = false`.

All the CPI instruments preserve their capability by using `CPI::InterpolationType` to let `CPICoupon` do the interpolation of the fixing.

So is this a problem for the `ZeroCouponInflationSwap` not being able to interpolate anymore? I would think so.

Do you see any possibility to solve this problem? Introducing an `InflationCashFlow` deriving from `IndexedCashFlow` which will do the interpolation like `CPICoupon` perhaps?

lballabio commented:
As for `InflationTermStructure:: indexIsInterpolated`, that was probably the wrong name for the method; I'm guessing the meaning would be more like `fixingsAreInterpolated`.  However, I'm not sure that this would be a property of the curve, once the dust settles on these changes.  Is this used in the code, and if so, how?

As you say, `ZeroCouponInflationSwap` should of course keep the possibility to use interpolated fixings.  An `InflationCashFlow` might be a good way to do so.  Before doing this, I'd also add a test case with `interpolated = true` so you can check the changes.

ralfkonrad commented:
> As you say, `ZeroCouponInflationSwap` should of course keep the possibility to use interpolated fixings. An `InflationCashFlow` might be a good way to do so. Before doing this, I'd also add a test case with `interpolated = true` so you can check the changes.

Agreed, I will focus on the `ZeroCouponInflationSwap` next.

The test suite is another part to discuss later on. Currently I have just `QL_DEPRECATED_DISABLE_WARNING_III` all warnings in there. A couple of testcases need attention where deprecated

https://github.com/lballabio/QuantLib/blob/9aa5cfb6b4a3be35ba0fd2e0fe1bd08ffc864e2f/test-suite/inflation.cpp#L228

is used.

> As for `InflationTermStructure:: indexIsInterpolated`, that was probably the wrong name for the method; I'm guessing the meaning would be more like `fixingsAreInterpolated`. However, I'm not sure that this would be a property of the curve, once the dust settles on these changes. Is this used in the code, and if so, how?

`fixingsAreInterpolated` is definitely what it means. That would have been a more obvious name for `bool InflationIndex::interpolated() const;` as well. It is not used that much as far as I can see at the moment. One key usage is

https://github.com/lballabio/QuantLib/blob/9aa5cfb6b4a3be35ba0fd2e0fe1bd08ffc864e2f/ql/termstructures/inflation/inflationtraits.hpp#L46-L53

It also pops up at here

https://github.com/lballabio/QuantLib/blob/9aa5cfb6b4a3be35ba0fd2e0fe1bd08ffc864e2f/ql/experimental/inflation/interpolatedyoyoptionletstripper.hpp#L171

but I think that is because `class YoYCapFloorTermPriceSurface` is derived  from `InflationTermStructure` rather than `YoYInflationTermStructure`.

It might be in more templated interpolators. They do not pop up in VS if there are not used (e.g. a test cases using them).

So there is still some dust to settle...

lballabio commented:
I see, thanks.  Yes, I guess you can deprecate `indexIsInterpolated` like you did for `interpolated`.

ralfkonrad commented:
> As you say, `ZeroCouponInflationSwap` should of course keep the possibility to use interpolated fixings. An `InflationCashFlow` might be a good way to do so. Before doing this, I'd also add a test case with `interpolated = true` so you can check the changes.

I will work on the `ZeroCouponInflationSwap` now in an extra PR as a precondition to this one.

> I see, thanks. Yes, I guess you can deprecate `indexIsInterpolated` like you did for `interpolated`.

And I think I would like to have another PR for the `InflationTermStructure:: indexIsInterpolated` once this is finished. Otherwise the changes are getting to "big" for me to keep track of them.

lballabio commented:
Sure.

lballabio commented:
@ralfkonrad  I merged master into this so it's up to date with the work on zero-coupon inflation swaps from last release.

What should the next step be?

ralfkonrad commented:
@lballabio I have to have a look into it again. I hope/think I can give you an answer during this weekend.

lballabio commented:
Sure, no problem

ralfkonrad commented:
@lballabio I have reverted the clang formatting changes so what we now have are the pure changes due to the deprecation of the interpolation feature.

Currently I see the following point we still have to work on

#### Within the library

- In `inflationindex.hpp` I have defined a couple of `QL_DEPRECATED...III...` macros as an extension to `QL_DEPRECATED`, `QL_DEPRECATED_DISABLE_WARNING` and `QL_DEPRECATED_ENABLE_WARNING` so we can turn on warning about the deprecated constructor(s), the `bool interpolated() const;` method and the `bool interpolated_ = false;` member separately. 
This made it easier for me to see the parts to focus on without getting mad about all the warnings. 
They should be removed before merging this PR and be replaced by the original once.

- The constructors using the `bool interpolated` argument are all deprecated and replaced by ones without the flag.

- The `bool interpolated_ = false;` member is only called in `inflationindex.cpp` and a couple of also deprecated index constructors like `UKRPI` etc. so that part is done.

- The `bool interpolated() const;` method is still used in the `CPICapFloorTermPriceSurface` constructor. The rest of the calls are removed and replaced by a `CPI::InterpolationType detail::CPI::effectiveInterpolationType(...)` call.

- The whole `InflationTermStructure` part needs attention. It still uses `bool indexIsInterpolated` which does not make sense anymore.

#### Test cases

- The test cases in `inflationzciisinterpolation.cpp` are not _real_ testcases yet. When we merged #1068 I was not aware that this was part of it. In `void runTest(const Setup& setup) { ... }` I just prepared the test cases but no actual test is done there. The idea for the test is "stolen" from the `InflationTest` and should be extended to the different combinations
```cpp
    static void asIndexNotInterpolated();
    static void asIndexInterpolated();
    static void flatNotInterpolated();
    static void flatInterpolated();
    static void linearNotInterpolated();
    static void linearInterpolated();
```

- The other inflation linked test cases are not changed yet. I have just disabled the deprecated warning there. E.g. in `InflationTest::testZeroTermStructure()` an interpolated `UKRPI` index is tested. I think some of the test cases can be kept but each one needs attention.

#### The next steps

- Perhaps first finish the test cases in `inflationzciisinterpolation.cpp` if we want to keep them. After the deprecation they can be removed anyhow.
- Make `CPICapFloorTermPriceSurface` independent of the index. As far as I can see passing the `Handle<ZeroInflationTermStructure>` taken from the index should be enough.
- Decide how to handle the `InflationTermStructure`.
- Fix the test case along the way.

The whole inflation implementation including instruments, indexes and term structures are highly dependent on each other. I think there is still a lot of work to do when we want to smoothly remove the index interpolation without breaking the current interface.

lballabio commented:
@ralfkonrad , this looks ready to me.  What do you think?

ralfkonrad commented:
:+1: Not sure though what's still my part in this PR :wink:

lballabio commented:
That's how it should be in a team effort, I guess :smile:
